### ISSUE TITLE
AppStoreClient does not work with new PuppetJs

### ISSUE THREAD

**Issue by miyconst** [Polyjuice/AppStoreClient#6]
AppStoreClient extracts db name from `puppet.referer` which is no longer exists. This could be easily fixed by replacing `puppet.referer` with `puppet.remoteUrl`.

The other issue, which I am not sure how to solve is `X-Referer` header.

Before:
```js
if (that.referer) {
    req.setRequestHeader('X-Referer', that.referer);
}
```

Now:
```js
if (that.remoteUrl) {
    req.setRequestHeader('X-Referer', that.remoteUrl);
}
```

The `referer` value was set after connection established, but the `remoteUrl` value is set from initial parameters.

AppStoreClient creates another puppet with 
```
remoteUrl = "http://localhost:8181/api/servermodel/default"
```

If the `X-Referer` set the `XMLHttpRequest` first sends `OPTIONS` instead of `GET` request. The server answers with `404` and the client fails.

```
XMLHttpRequest cannot load http://localhost:8181/api/servermodel/default. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://localhost:8080' is therefore not allowed access. The response had HTTP status code 404.
```

**Comment by miyconst** [comment]
@warpech  and me agreed to:

1. Set `X-Referer` header only if the `url` of the outgoing request is not the same as header value.
2. Save `this.remoteUrl` as `URL` object, not a string.
3. Inside the `setRemoteUrl` function calculate `remoteUrl` as `new URL(remoteUrl, this.remoteUrl.href)`.
4. Set `X-Referer` header value like `this.remoteUrl.path`.

**Comment by warpech** [comment]
On second thought, controlling this using a new boolean argument to the `xhr` method, as you proposed on Skype, also doesn't look that bad and is even clearer.

**Comment by miyconst** [comment]
Okay, I will add the argument then.

I would like to take this issue after this https://github.com/PuppetJs/PuppetJs/issues/55 is merged,
could you or @tomalec check it please?

**Comment by warpech** [comment]
Please merge https://github.com/PuppetJs/PuppetJs/issues/55 if it's stopping you

**Comment by miyconst** [comment]
Another problem appears.

`http://localhost:8181/` refuses PuppetJs cross domain requests.

After recent changes PuppetJs doesn't wait till WebSocket connection opens, but sends requests over http.

AppStoreClient loads applications by 
```js
puppet.obj.RefreshAppStoreStores$++;
```

This causes an http request to

```
http://localhost:8181/api/servermodel/default/4f0389c9-e21e-452e-8596-f6acaf62ffdd/39F2F8DE0E29532B10000010 
```

Which is refused.

The easiest way to solve it, is to load applications in timeout.

**Comment by tomalec** [comment]
Could we use some kind of connectionEstablished event instead of timeout?

**Comment by miyconst** [comment]
There is already an event `onSocketStateChanged`, so I can use it.

**Comment by warpech** [comment]
I think solving it server side would be even better. Access-Control headers should be added to https://github.com/Starcounter/Starcounter/blob/6409aba96274e7059d56a0b2b07d754f2f40daa8/src/Starcounter.Administrator/Server/Managers/ServerManager.cs#L80

@alemoi could you do that?

**Comment by tomalec** [comment]
:+1:
Probably, you will also need to make sure that it works for HTTP connections, and WS that opens pretty fast ;P

**Comment by miyconst** [comment]
>Probably, you will also need to make sure that it works for HTTP connections, and WS that opens pretty fast ;P

HTTP does not work for now, and WS connection which opens very fast works.

**Comment by alemoi** [comment]
@warpech I can take a look, just describe what exactly should be done there, I didn't really get the idea.

**Comment by warpech** [comment]
@miyconst will explain :)

**Comment by miyconst** [comment]
@alemoi, cross domain javascript requests are blocked by default. 

When a browser detects a corss domain javascript request it first sends `OPTIONS` request to the same url and expects a response with `Access-Control-Allow-Origin` header. If the value of the header allows requests from current host, then the normal `GET`/`POST` request being send.
http://enable-cors.org/server_aspnet.html

In our case you should add a line of code:
```cs
Handle.AddOutgoingHeader("Access-Control-Allow-Origin", "*");
```

This will allow cross domain requests from every host.

**Comment by alemoi** [comment]
Sorry for the late response. Should the suggested line be only added to handler: https://github.com/Starcounter/Starcounter/blob/6409aba96274e7059d56a0b2b07d754f2f40daa8/src/Starcounter.Administrator/Server/Managers/ServerManager.cs#L80
or somewhere else?

**Comment by miyconst** [comment]
The AppStoreClient (and me) right now needs it only inside the handler. 

You can also use the old syntax from the previous handlers:

```js
response["Access-Control-Allow-Origin"] = "*";  //"http://localhost:8080";
```

**Comment by alemoi** [comment]
Fixed via 3d0c9b5e8ff9cf6d9ba82dcde1af050fd01a563d

**Comment by miyconst** [comment]
It seems like @alemoi has done what we asked for, but it still does not work.
Server responses with `404` to `OPTIONS` request.

The solution should be:
```cs
Handle.OPTIONS("/api/servermodel/{?}/{?}", (string id, Session session, Request request) => {
    response["Access-Control-Allow-Origin"] = "*";  //"http://localhost:8080";
    response.StatusCode = (UInt16) System.Net.HttpStatusCode.OK;
    return response;
});
```

But `Starcounter.Handle` does not have `OPTIONS` method.

**Comment by alemoi** [comment]
> But Starcounter.Handle does not have OPTIONS method.

Handle.CUSTOM can be used, like this: `Handle.CUSTOM("OPTIONS /myuriwithparameters", ...);`
But I will add support for OPTIONS method now.

**Comment by warpech** [comment]
This is so low level... 

In ExpressJS and Meteor, CORS is enabled on any HTTP method by using a simple filter:

```
app.use(function(req, res, next) {
  res.header("Access-Control-Allow-Origin", "*");
  res.header("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept");
  next();
});
//source: http://enable-cors.org/server_expressjs.html
```

In Koa it is the same, but abstracted away as a module:

```
var cors = require('koa-cors');
var app = koa();
app.use(cors());
//source: https://github.com/evert0n/koa-cors
```

@miyconst could you check if we can achieve the same using `AddFilterToMiddleware`? I assume `AddFilterToMiddleware` runs for all HTTP methods.

**Comment by miyconst** [comment]
Yes, the following filter adds header to every request:

```cs
Handle.AddFilterToMiddleware((Request) => {
    Handle.AddOutgoingHeader("Access-Control-Allow-Origin", "*");

    return null;
});
```

But if the request was not handled, then no filters were applied, so, `Handle.OPTIONS` is still required.

I have tested `Handle.OPTIONS` and it works fine, but the url: `http://localhost:8181/api/servermodel/default/d5a9b1bf-4721-44cf-8f36-4844bf81f2c2/507B0DBDBB056D1C00000040` still returns `404`. I guess me and @alemoi  should sit down and figure out a solution.

**Comment by miyconst** [comment]
The issue is fixed now.

**Comment by warpech** [comment]
:point_up_2:  I created a new issue about using filters to setup CORS.
